cabal-install/Distribution/Client/CmdSdist.hs

```diff
@@ -255,10 +256,11 @@ packageToSdist verbosity projectRootDir format outputFile pkg = do
                             Right path -> tell [Tar.directoryEntry path]
 
                         for_ files $ \(perm, file) -> do
+                            realPerm <- getPermissions file
                             let fileDir = takeDirectory (prefix </> file)
                                 perm' = case perm of
-                                    Exec -> Tar.executableFilePermissions
-                                    NoExec -> Tar.ordinaryFilePermissions
+                                    Exec | executable realPerm -> Tar.executableFilePermissions
```

Comment:
It seems rather odd that we can have files marked as `Exec` that don't have executable permissions. Would it not be cleaner to change the implementation of `listPackageSources` so that these files are marked as `NoExec` instead?

Reply:
On reflection, that's weird. I think, instead of surgically fixing this issue, it might be worth considering a more thorough change to properly read the permissions and set it based on that, and not bother using the cruddy heuristic that we currently have at all, unless there's an important reason I'm not remembering for why we want to ignore some files' permissions.

This does have the downside that changing permissions (and not file contents) will now be able to invalidate package hashes, which I think is the original reason I had it not check.

Reply:
The culrpit looks to be in `SrcDist.hs`

```haskell
-- | List those source files that may be executable (e.g. the configure script).
listPackageSourcesMaybeExecutable :: Verbosity -> PackageDescription -> IO [FilePath]
listPackageSourcesMaybeExecutable verbosity pkg_descr =
  -- Extra source files.
  fmap concat . for (extraSrcFiles pkg_descr) $ \fpath ->
    matchDirFileGlob verbosity (specVersion pkg_descr) "." fpath
```

i.e. every `extraSrcFiles` are given executable bits now? e.g. for `lens-4.18.1.tar.gz` from Hackage:

```haskell
% tar -tzvf lens-4.18.1.tar.gz|grep README
-rwxr-xr-x 0/0            7427 2001-09-09 04:46 lens-4.18.1/README.markdown
```

IMHO that's just plain wrong. 

What would break if *all* files are `644` (and directories 755)?

For example, `configure` script is actually run by invoking `sh` on it, see `runConfigureScript`

```haskell
  shConfiguredProg <- lookupProgram shProg
                      `fmap` configureProgram  verbosity shProg progDb
  case shConfiguredProg of
      Just sh -> runProgramInvocation verbosity $
                 (programInvocation (sh {programOverrideEnv = overEnv}) args')
                 { progInvokeCwd = Just (buildDir lbi) }
      Nothing -> die' verbosity notFoundMsg
```

(and https://github.com/haskell/cabal/issues/5813#issuecomment-564981835 suggest that if we unpack without executable bits, why store them in the first place).

I would be worried of invalidating cache. It's cache for a reason.